### DISCUSSION TITLE
Compiling with SoftwareSerial.h not possible (ATtiny2313, memory problem)

### DISCUSSION THREAD

**Discussion by Gehilfe** [SpenceKonde/ATTinyCore#518] [Support / Q & A]
Hi,
I am trying to compile a very simple program with the library SoftwareSerial.h for an ATtiny2313 and I get the error message directly:

```
Sketch uses 2344 bytes (114%) of program storage space. Maximum is 2048 bytes.text section exceeds available space in board
Global variables use 144 bytes (112%) of dynamic memory, leaving -16 bytes for local variables. Maximum is 128 bytes.
```

The program code looks like this:
```
#include <SoftwareSerial.h>

SoftwareSerial device(3,4); //Rx,Tx
byte stunden = 0;
byte minuten = 0;
byte sekunden = 0;
byte startcode = B10101010;
byte empfangenercode = 0;

void setup() {
  device.begin(9600);
  device.println("Setup erledigt");
}

void loop() {
  if (device.available() > 0) {
    //Serial.println("Daten vorhanden");
    empfangenercode = device.read();
    device.println(empfangenercode,BIN);
    if (empfangenercode == startcode) {
      stunden = device.read();
      minuten = device.read();
      sekunden = device.read();
      device.println(stunden);
      device.println(minuten);
      device.println(sekunden);
    }
  }
}
```
I use the Arduino IDE 1.8.13 and the ATTinyCore 1.4.1. 
The same problem occurs when I try to compile the program with PlatformIO.

Am I doing something wrong? Do you have a tip for me? For the final program (a multiplexer for Nixie tubes) I only need the "serial read function", but for debugging it would be great for me to be able to read as well as write serially.

Would be great if you could help me.

**Comment by SpenceKonde** [comment]
Why software serial??? There is a real hardware serial port just sitting there.. hardware serial is generally less resource hungry than software serial and works better. That alone might save you enough to get in under the limit,, check also 
You can (and should) wrap any debugging strings in F() macro, but that only saves ram, whereas you're barely over the RAM limit, but way over the flash limit. I think printing strings is a luxury you can't afford . Maybe print char's (single quotes, like 'a', take 1 byte of storage; it takes 2 bytes to store "a" vs 1 t store 'a' because the one with double quotes is a null terminated char array,with 2 elements 

```
byte stunden = 0;
byte minuten = 0;
byte sekunden = 0;
byte startcode = B10101010;
byte empfangenercode = 0;

void setup() {
  Serial.begin(9600);
  //Serial.println("Setup erledigt");
  //Serial.prrintln 
}

void loop() {
  if (Serial.available() > 0) {
    //Serial.println("Daten vorhanden");
    empfangenercode = Serial.read();
    Serial.println(empfangenercode,BIN);
    if (empfangenercode == startcode) {
      stunden = Serial.read();
      minuten = Serial.read();
      sekunden = Serial.read();
      Serial.println(stunden);
      Serial.println(minuten);
      Serial.println(sekunden);
    }
  }
}
```

Anyway, my other tip would be to use a part with more than 2k of flash. The 2313 is the lamest part supported by this core (okay, maybe the 24 is worse). I am amazed how many people are still using these dreadful ancient parts. They aren't even particularly cheap! digikey lists them for like 75 cents/ea in cheapest package that can be soldereed without advanced techniques, no minimum order, which would get you into 8k flash tinyAVR 0/1-series parts, and it's very close to the price of the stand-out ATtiny1614, with 16kl flash, 2k ram, and the "good set" of tinyAVR 1-series peripherals (the 16k and 32k 1-series have a second ADC, second type B timer, and some other goodies that I can't remember off hand atm).

**Reply by Gehilfe** [reply]
Thank you for your prompt reply. Maybe I should really switch to another Tiny. For me it was the first attempt with the ATtiny. I wanted to use the SoftwareSerial connection in addition to the regular connection to receive data from another controller while sending other data to the SoftwareSerial for debugging (but only during development). The problem was that I could never receive anything with the FTDI breakout board with the regular Serial.println (). I will test it again this afternoon with a "fresh" ATtiny2313 and order a somewhat "larger" model at the same time. Is the ATtiny1614 good for that?

**Reply by SpenceKonde** [reply]
The ATTiny1614 is a very different part - but it is a spectacularly capable one! Supported by my megaTinyCore (which is also in a lot of ways a better core - I was more experienced when I was making key design decisions), it has a bunch of the latest features, and (because it's a 16k+ 1-series) it's got the "cadillac" features - second type B timer, whole second ADC (yeah, seriously - not terribly useful with analogRead(), but if you're throwing them into free-running mode, you can do unreal things with it), hardware serial port with fractional baud rate generator, the internal oscillator is very accurate.... generally a wonderful part. You know how a lot of people default to reaching for an atmega328p? **_I generally default an attiny1614 these days!. (_** (assuming I don't need multiple serial ports and I'm not trying to abuse heck out of the CCL, both of which call for an AVR128Dx
